### DISCUSSION TITLE
Quicker ways to combine two images in Magick.net

### DISCUSSION THREAD

**Discussion by herbertgzb** [dlemstra/Magick.NET#827] [Help]
The task I am going through is rather straightforward -- I need to read either a GROUP IV tiff image or a jpeg image. I then need to either cut a slice in the that image or to annotate it. Either way -- using magick.net I would think the first task is to cut the image into two pieces (top and bottom) and cut out the middle part, and combine them. The second task would be to create a canvas and write on that canvas with the original image and the annotation.

A sample code for the first task would be like this (after I have 
```

'read in image
Dim image_top As New MagickImage(inputpath.ToString)
 Dim image_bottom As New MagickImage(inputpath.ToString)

'crop top slice and bottom slice with specific heights
image_top.Crop(image_top.Width, image_top.Height - CInt(centerslice_height.Text) - CInt(bottomslice_height.Text), Gravity.North)
image_bottom.Crop(image_bottom.Width, CInt(bottomslice_height.Text), Gravity.South)

'Combine top and bottom slices
Dim output As New MagickImageCollection
output.Add(image_top)
output.Add(image_bottom)
Dim IMAGE As New MagickImage(output.AppendVertically)
IMAGE.Format = MagickFormat.Tiff
IMAGE.Settings.Compression = 8
IMAGE.Write(outputpath)

```
Up to the "IMAGE.Write(outputpath)" line, everything took around 80ms. The final line takes around 1500ms/1.5s either writing on network share or locally.

If the task is simpler/not involving combining the two images together, aka, like below

```
Dim image_top As New MagickImage(inputpath.ToString)
image_top.Crop(image_top.Width, image_top.Height - CInt(centerslice_height.Text) - CInt(bottomslice_height.Text), Gravity.North)
image_top.Write(outputpath)

```
The above code takes around 80ms to finish. Note that -- the image_top contains most of the whole image (image top is around 3100 pixels while image_bottom is around 100 pixels). So I don't think size itself is the problem here.

I have also tried the composite method (I believe for annotation I'll use the same method)
```

'read in image
Dim image_top As New MagickImage(inputpath.ToString)
Dim image_bottom As New MagickImage(inputpath.ToString)

'crop top slice and bottom slice with specific heights
image_top.Crop(image_top.Width, image_top.Height - CInt(centerslice_height.Text) - CInt(bottomslice_height.Text), Gravity.North)
image_bottom.Crop(image_bottom.Width, CInt(bottomslice_height.Text), Gravity.South)

'use a canvas to write top/bottom slices
 Dim settings As New MagickReadSettings()
settings.Width = image_top.Width
settings.Height = image_top.Height + image_bottom.Height
settings.Compression = 8
Dim IMAGE = New MagickImage(outputpath, settings)
IMAGE.Format = MagickFormat.Tiff
IMAGE.Composite(image_top, 0, 0)
IMAGE.Composite(image_bottom, image_top.Height, 0)
IMAGE.Write(outputpath)

```
And again -- the last step take 1.5s vs everything before the IMAGE.Write(outputpath) line taking less than 80ms.

Wonder if I can do anything to make the whole process quicker?

Thanks in advance!

**Comment by dlemstra** [comment]
Without your images I cannot really give you a good answer. I suspect that the compression is what is making it take the 1.5s on your machine.

**Reply by herbertgzb** [reply]
Hey -- thank you. Sorry for revisiting this after so long, being distracted.

It is amazing that you actually hold up two projects -- both imagemagick and GraphicsMagick (as I understand? I believe you also show up in the imagesharp project as well). Since the stuff I am trying to do does not require a lot of in depth/complex manipulations but really, cropping images, stitching images, converting jpgs to tiffs and vice versa -- I used the graphicsmagick library and somehow I think I am yielding a lot faster result.

My question back to you -- the graphicsmagick requires c++ redistributable 2012 (whereas I have 2015-2019 so I had to install that). Does the built-application also need c++ redistributable 2012 on the machines I want the application to run?

Thank you again!

**Reply by dlemstra** [reply]
The GraphicsMagick.NET library is no longer being maintained. I do think you need to install that redistributable to on the machines that run your application.

**Reply by herbertgzb** [reply]
> The GraphicsMagick.NET library is no longer being maintained. I do think you need to install that redistributable to on the machines that run your application.

Thanks for answering. I literally used the exact same code whereas the graphicsmagic is about 10 times faster. It does look as if there's a memory leak after 20 minutes, even if I disposed them (had to use GC.collect though I guess I shouldn't?). but since that code is no longer maintained, I guess I'll live with it.

Anyways, thanks a lot for maintaining such amazing projects! It does interest me that neither imagesharp nor a lot of the imaging processing libraries maintain a good group IV tiff support. It looks as if people are doing web dev mostly and group IV tiff is not a favorite in those scenarios. Are we expecting imagesharp to support group IV tiff any time soon? Any other good libraries that you'd suggest? I know leadcode and atalasoft's dotimage libraries can do some heavy work but I am not using my applications to a commercial use/for profit, etc...

**Reply by dlemstra** [reply]
It might be possible that GraphicsMagick is using a different form of compression. There is working being done on adding support for TIFF to ImageSharp here: https://github.com/SixLabors/ImageSharp/pull/1553. I don't know about any other libraries that offer tiff support. Have you tried this library: https://github.com/BitMiracle/libtiff.net? This is build on top of libtiff that is also used by Magick.NET and GraphicksMagick.NET

If it would be possible to share an image and code sample that includes the minimum about of code I might be able to give a better answer on why this takes a long time to execute.

**Reply by herbertgzb** [reply]
Thanks again for commenting! Yes -- on the imageshark part, I looked at that and also another older pull request,both of which said, Group IV is not supported as of current.

I in fact also checked out the OPENCV and its .net wrapper EMGUCV, but I think they apply in different aspects of imaging processing. I am more looking for a "read an image, crop an image, outputs an image" library, whereas OPENCV probably gets to the bottom of it as it scans line by line pixel by pixel. Same as what I read from libtiff. Being a side project I am not sure if I am going to do that...Sorry for the laziness! 


I'll do a separate post on the code and try to share the actual image. The image is a 2550x3300 300 dpi group IV image. I am trying to slice it into 3 parts and cut off the center part and stitch the top/bottom part together, basically. I probably also would later try to extend the image onto a slightly larger "canvas" and add text in the extra part. Imagemagick/every other library are pretty quick if I simply try to crop the image, but starts to get real slow when I apply the larger canvas/extend it, etc.

**Reply by dlemstra** [reply]
Are you using the 64-bit build? Maybe you are running into 32-bit memory limits?

**Reply by herbertgzb** [reply]
Image magick code


        'Read image
        Dim IMAGE_TOP As New MagickImage(Row("inputpath").ToString)
        Dim image_bottom As New MagickImage(Row("inputpath").ToString)

        'crop top slice (height =  total - retain - crop) and bottom slice (height = NBox_Ret_Bottom), leave out middle slice (height = Nbox_Crop_Bottom)
        IMAGE_TOP.Crop(IMAGE_TOP.Width, IMAGE_TOP.Height - NBox_Ret_Bottom.Value - Nbox_Crop_Bottom.Value, Gravity.North)
        image_bottom.Crop(image_bottom.Width, NBox_Ret_Bottom.Value, Gravity.South)

        'stitch top and bottom
        Dim output As New MagickImageCollection
        output.Add(IMAGE_TOP)
        output.Add(image_bottom)
        Dim IMAGE As New MagickImage(output.AppendVertically)

        'output
        If IMAGE.Depth <> 1 Then
            outputpath = outputpath & ".jpg"
            IMAGE.Format = MagickFormat.Jpg
            IMAGE.Settings.Compression = 12
        Else
            outputpath = outputpath & ".tif"
            IMAGE.Format = MagickFormat.Tiff
            IMAGE.Settings.Compression = 8

        End If

        IMAGE.Write(outputpath)

        IMAGE.Dispose()
        image_bottom.Dispose()
        IMAGE_TOP.Dispose()
        output.Dispose()
`

this takes about 1.5s for the sample image.

For graphics magick the code is almost identical, except that it looks as if I need to force a GC.Collect at the end of each image it processes else there's a memory leak after a while. For same image it takes about 150ms



        'Read image
           Dim IMAGE_TOP As New MagickImage(ROW("inputpath").ToString)
            Dim image_bottom As New MagickImage(ROW("inputpath").ToString)
         'Crop top slice
            IMAGE_TOP.Crop(IMAGE_TOP.Width, IMAGE_TOP.Height - NBox_Ret_Bottom.Value - Nbox_Crop_Bottom.Value, Gravity.North)
            image_bottom.Crop(image_bottom.Width, NBox_Ret_Bottom.Value, Gravity.South)
       'Stitch top & bottom
            Dim output As New MagickImageCollection
            output.Add(IMAGE_TOP)
            output.Add(image_bottom)
            Dim IMAGE As New MagickImage(output.AppendVertically)
            If image.Depth <> 1 Then
                outputpath = outputpath & ".jpg"
                image.Format = MagickFormat.Jpg
                IMAGE.CompressionMethod = CompressionMethod.JPEG
            Else
                outputpath = outputpath & ".tif"
                image.Format = MagickFormat.Tiff
                IMAGE.CompressionMethod = CompressionMethod.Group4
            End If
            image.Write(outputpath)
            image.Dispose()
            image_bottom.Dispose()
            IMAGE_TOP.Dispose()
            output.Dispose()
            System.GC.Collect()
`

Sample image can be found attached -- it don't support tiff so had to upload a converted jpeg...you can convert it back but I believe it's either way
![KPLUSS-02869621](https://user-images.githubusercontent.com/76821472/109714457-70907100-7b70-11eb-9ad7-04fe363bceac.jpg)

**Reply by herbertgzb** [reply]
> Are you using the 64-bit build? Maybe you are running into 32-bit memory limits?

For this part I might need to test on that. I am using the 64-bit build.

**Reply by dlemstra** [reply]
It turns out that you helped me find an optimization inside Magick.NET. I would advise you to use `image.DetermineColorType();` after I have published a new version of ImageMagick. This will change the type of this image to `Bilevel` instead of doing an expensive `Quantize` operation inside the `TIFF` encoder.

And I will also try to figure out why the type of this image was changed when doing the `Crop` and `AppendVertically` operation. It could remain `Bilevel` instead. But I am not sure if I can make this change,

**Reply by herbertgzb** [reply]
> It turns out that you helped me find an optimization inside Magick.NET. I would advise you to use `image.DetermineColorType();` after I have published a new version of ImageMagick. This will change the type of this image to `Bilevel` instead of doing an expensive `Quantize` operation inside the `TIFF` encoder.
> 
> And I will also try to figure out why the type of this image was changed when doing the `Crop` and `AppendVertically` operation. It could remain `Bilevel` instead. But I am not sure if I can make this change,

Great to know and thanks! 
To clarify -- I should do this before I do the crop, right? And also before the append vertically? But the appendvertically is done the same time as I define the object?

**Reply by dlemstra** [reply]
Sorry for the late response @herbertgzb. The  `DetermineColorType` operation should indeed be the first one. I also added another optimization in the latest version of Magick.NET that will keep the image as `Bilevel` in the call to `AppendVertically`. That should improve the performance even more.

**Reply by herbertgzb** [reply]
> Sorry for the late response @herbertgzb. The `DetermineColorType` operation should indeed be the first one. I also added another optimization in the latest version of Magick.NET that will keep the image as `Bilevel` in the call to `AppendVertically`. That should improve the performance even more.

Thanks a lot! I'll definitely check out the latest release!!
